### PR TITLE
Mark shares as dirty when changing permissions

### LINKED ISSUE #16587: Changing share permissions should trigger an etag change
As discussed with @guruz @ogoffart @PVince81: Whenever file metadata (mtime, permissions, ...) changes, the parent folder's etag should change. That notifies sync clients and allows them to check for changes.

The particular issue here is that changing the permissions of a share (readonly<->readwrite) does not trigger an etag change. Therefore the sync client does not notice that the permissions of the shared files have changed.

Required for owncloud/client#3244

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #16587

cc @PVince81 @DeepDiver1975 @MorrisJobke

### CODE REVIEW

**Comment by karlitschek** [comment]
nice :+1:

**Comment by DeepDiver1975** [comment]
@icewind1991 milestone? labels? :imp: 

and there is a unit test failing - THX

**Comment by icewind1991** [comment]
Unit tests pass locally...

**Comment by PVince81** [comment]
Cool, so looking at the code it seems it was easy than I thought.

**Comment by DeepDiver1975** [comment]
> Unit tests pass locally...

not for me against sqlite

````
/home/deepdiver/Development/ownCloud/master-autotest/tests/lib/mail/message.php:45

4) OCA\Files_sharing\Tests\EtagPropagation::testEtagChangeOnPermissionsChange
Failed asserting that the etag for "" of user test-share-user2 has changed
Failed asserting that '560d51b04f494' is not equal to <string:560d51b04f494>.

/home/deepdiver/Development/ownCloud/master-autotest/apps/files_sharing/tests/etagpropagation.php:137
/home/deepdiver/Development/ownCloud/master-autotest/apps/files_sharing/tests/etagpropagation.php:166
/home/deepdiver/Development/ownCloud/master-autotest/apps/files_sharing/tests/etagpropagation.php:420

**Comment by PVince81** [comment]
Tested manually, worked.

Local unit tests with sqlite also pass for me. Race condition ?

**Comment by rullzer** [comment]
Also working here locally

**Comment by PVince81** [comment]
@DeepDiver1975 can you try again ? Hopefully the rebase helped...

**Comment by MorrisJobke** [comment]
```
OCA\Files_sharing\Tests\EtagPropagation::testEtagChangeOnPermissionsChange
Failed asserting that the etag for "" of user test-share-user2 has changed
Failed asserting that '560e8fbe92a05' is not equal to <string:560e8fbe92a05>.

/var/jenkins/workspace/core-ci-linux@3/database/oci/label/SLAVE/apps/files_sharing/tests/etagpropagation.php:137
/var/jenkins/workspace/core-ci-linux@3/database/oci/label/SLAVE/apps/files_sharing/tests/etagpropagation.php:166
/var/jenkins/workspace/core-ci-linux@3/database/oci/label/SLAVE/apps/files_sharing/tests/etagpropagation.php:420
```

also the php7 job fails

**Comment by MorrisJobke** [comment]
@PVince81 It's only on OCi ;)

**Comment by DeepDiver1975** [comment]
This still falls apart - even if it work on other machines - it's fishy and we cannot accept this.

@icewind1991 please try to find a solution - THX

**Comment by icewind1991** [comment]
Fixed the race condition

**Comment by DeepDiver1975** [comment]
> Fixed the race condition

great - THX

**Comment by DeepDiver1975** [comment]
@PVince81 @MorrisJobke @rullzer please review and test - THX

**Comment by MorrisJobke** [comment]
This only changes the etag of the parent if I do a PROPFIND against the parent. I think this is intended. Shouldn't there be a permission field in WebDAV? I couldn't find it. I guess I need to specify that I also want the permissions for each element.

Also nothing in the log :+1: from my side.

File: apps/files_sharing/lib/propagation/changewatcher.php [hunk 1]
```diff
@@ -99,4 +99,12 @@ private function propagateForOwner($share, $internalPath, ChangePropagator $prop
 			$propagator->propagateChanges();
 		}
 	}
+
+	public function permissionsHook($params) {
+		$share = $params['share'];
+
+		if ($share['item_type'] === 'file' || $share['item_type'] === 'folder') {
+			$this->recipientPropagator->markDirty($share, microtime(true));
```
Comment (rullzer) [inline apps/files_sharing/lib/propagation/changewatcher.php]: this break the function signature... https://github.com/owncloud/core/blob/etag-share-permissions-update/apps/files_sharing/lib/propagation/recipientpropagator.php#L106-L110

**Comment by rullzer** [comment]
Other than my comments. Looking good and seems to work! :+1:

File: apps/files_sharing/lib/propagation/changewatcher.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_sharing/lib/propagation/changewatcher.php]: phpdoc is wrong, argument should be float, will fix

**Pull request merged**